common/quotas/dynamicratelimiter.go

```diff
@@ -24,48 +24,103 @@ package quotas
 
 import (
 	"context"
+	"math"
+	"sync/atomic"
+	"time"
 
 	"golang.org/x/time/rate"
 
 	"github.com/uber/cadence/common/clock"
 )
 
+const _ttl = time.Second * 5
+const _minBurst = 1
+
 // DynamicRateLimiter implements a dynamic config wrapper around the rate limiter,
 // checks for updates to the dynamic config and updates the rate limiter accordingly
 type DynamicRateLimiter struct {
-	rps RPSFunc
-	rl  *RateLimiter
+	rps            RPSFunc
+	rl             clock.Ratelimiter
+	timeSource     clock.TimeSource
+	ttl            time.Duration
+	lastUpdateTime atomic.Value
+	minBurst       int
+}
+
+type DynamicRateLimiterOpt = func(limiter *DynamicRateLimiter)
+
+func WithTTL(ttl time.Duration) DynamicRateLimiterOpt {
+	return func(limiter *DynamicRateLimiter) {
+		limiter.ttl = ttl
+	}
+}
+
+func WithMinBurst(minBurst int) DynamicRateLimiterOpt {
+	return func(limiter *DynamicRateLimiter) {
+		limiter.minBurst = minBurst
+	}
+}
+
+func WithTimeSource(timeSource clock.TimeSource) DynamicRateLimiterOpt {
+	return func(limiter *DynamicRateLimiter) {
+		limiter.timeSource = timeSource
+	}
 }
 
 // NewDynamicRateLimiter returns a rate limiter which handles dynamic config
-func NewDynamicRateLimiter(rps RPSFunc) *DynamicRateLimiter {
-	initialRps := rps()
-	rl := NewRateLimiter(&initialRps, _defaultRPSTTL, _burstSize)
-	return &DynamicRateLimiter{rps, rl}
+func NewDynamicRateLimiter(rps RPSFunc, opts ...DynamicRateLimiterOpt) Limiter {
+	res := &DynamicRateLimiter{
+		rps:        rps,
+		timeSource: clock.NewRealTimeSource(),
+		ttl:        _ttl,
+		minBurst:   _minBurst,
+	}
+	for _, opt := range opts {
+		opt(res)
+	}
+	res.lastUpdateTime.Store(res.timeSource.Now())
+	res.rl = clock.NewRatelimiter(res.getLimitAndBurst())
+	return res
 }
 
 // Allow immediately returns with true or false indicating if a rate limit
 // token is available or not
 func (d *DynamicRateLimiter) Allow() bool {
-	rps := d.rps()
-	d.rl.UpdateMaxDispatch(&rps)
+	d.maybeRefreshRps()
 	return d.rl.Allow()
 }
 
 // Wait waits up till deadline for a rate limit token
 func (d *DynamicRateLimiter) Wait(ctx context.Context) error {
-	rps := d.rps()
-	d.rl.UpdateMaxDispatch(&rps)
+	d.maybeRefreshRps()
 	return d.rl.Wait(ctx)
 }
 
 // Reserve reserves a rate limit token
 func (d *DynamicRateLimiter) Reserve() clock.Reservation {
-	rps := d.rps()
-	d.rl.UpdateMaxDispatch(&rps)
+	d.maybeRefreshRps()
 	return d.rl.Reserve()
 }
 
 func (d *DynamicRateLimiter) Limit() rate.Limit {
-	return rate.Limit(d.rps())
+	d.maybeRefreshRps()
+	return d.rl.Limit()
+}
+
+func (d *DynamicRateLimiter) maybeRefreshRps() {
+	now := d.timeSource.Now()
+	lastUpdated := d.lastUpdateTime.Load().(time.Time)
+	if now.After(lastUpdated.Add(d.ttl-1)) && d.lastUpdateTime.CompareAndSwap(lastUpdated, now) {
```

Comment:
tbh I'd just drop the `-1` for simplicity, I doubt we care about nanosecond accuracy :)  and I think that'd mean your test-sleeps can become `.Advance(ttl)` too

Reply:
I like the `-1` because it allows for setting a ttl of `0` which fully disables this behavior. Otherwise you'd need to set the TTL to `-1` which seems counter intuitive.

Reply:
yea, 0 is a pretty reasonable value for that 🤔  sounds good enough to me.
(I guess the alternative would be to switch to `now.Compare(lastUpdated.Add(d.ttl)) > 0`?  that's not terrible, but also not obviously worth it, and it wouldn't change the -1/+1 fiddling in tests)